### PR TITLE
Code refactoring

### PR DESCRIPTION
**Pull request by Sidd-kp**
### Description

Code Refactoring and removal of small bugs. 
- Key Policies :
  - Adding of `flex.ResourceControllerURL` in schema which is later being set in the read function.
  - Fixing `TestAccIBMKmsDataSourceKeyPolicy_basicNew`
  - Remove `resourceIBMKmsKeyRead(d, meta)` from update function when there is error in policy creation
- Key Rings :
  - /ibm/service/kms/resource_ibm_kms_key_rings.go : line 92 :

> Might cause nil pointer dereference if `instanceData` is nil
```
instanceData, resp, err := rsConClient.GetResourceInstance(&resourceInstanceGet)
instanceCRN := instanceData.CRN
if err != nil || instanceData == nil {
     return fmt.Errorf("[ERROR] Error retrieving resource instance: %s with resp code: %s", err, resp)
}
```

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

**Comment by Sidd-kp** [comment]
Screenshot 2022-09-05 at 9 20 51 PM
as

**Comment by Sidd-kp** [comment]
Screenshot 2022-09-06 at 11 44 49 AM

**Comment by harshit777** [comment]
Can u provide the eveidence of resource provisioning using `terraform apply` ?

**Comment by harshit777** [comment]
Can u add more test case result evidences

**Comment by Sidd-kp** [comment]
Screenshot 2022-09-14 at 1 05 51 PM

File: ibm/service/kms/resource_ibm_kms_key.go [hunk 1]
```diff
@@ -403,59 +297,117 @@ func resourceIBMKmsKeyDelete(d *schema.ResourceData, meta interface{}) error {
...
 func KmsEndpointURL(kpAPI *kp.Client, endpointType string, extensions map[string]interface{}) (*url.URL, error) {
 
 	exturl := extensions["endpoints"].(map[string]interface{})["public"]
 	if endpointType == "private" || strings.Contains(kpAPI.Config.BaseURL, "private") {
 		exturl = extensions["endpoints"].(map[string]interface{})["private"]
 	}
-	endpointURL := fmt.Sprintf("%s/api/v2/keys", exturl.(string))
 
-	url1 := conns.EnvFallBack([]string{"IBMCLOUD_KP_API_ENDPOINT"}, endpointURL)
+	url1 := fmt.Sprintf("%s/api/v2/keys", conns.EnvFallBack([]string{"IBMCLOUD_KP_API_ENDPOINT"}, exturl.(string)))
```
Comment (kavya498) [inline ibm/service/kms/resource_ibm_kms_key.go:410]: This ll be a breaking change to existing users.. 
Please revert this change
